Distributed Fast Fourier Transforms

ClaudiaComito opened the pull request:
Resolves #1097  and can be used for #1248 

## Due Diligence
<!--- Please address the following points before setting your PR "ready for review".
--->
- General:
    - [x]  **base branch** must be `main` for new features, latest release branch (e.g. `release/1.3.x`) for bug fixes
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description

<!--- Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.
--->

Introducing `fft` module with all relevant operations.

Main points:
- Transforms along the split dimension are handled by redistributing to a different axis first, transforming the original split dimension locally, and then redistributing back.
- Multi-axis transforms on real input (`rfftn`, `ihfft` etc.) will be performed as the corresponding generic FFT operation  if transform of the split axis is required. This is due to the intrinsic recursion built into these operations (transpose, resplit, transform, resplit, transpose, transform), where the complex output of the first real transform cannot act as input to the second real transform.
- Multi-axis transforms on real input  still return the correct shape, i.e. for the last transformed axis only up to the Nyquist frequency.

Still TODO:

- [x] expand documentation esp. of real ffts
- [x] implement `fftfreq`, `fftshift` etc.
- [x] fix test failure (on 7 processes) EDIT: solved by setting correct output dtype of empty local tensor)




Issue/s resolved: #1097 

## Changes proposed:

- implemented FFT functions
- implemented tests
-
-

## Type of change
<!--
i.e.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation update
--->
- New feature (non-breaking change which adds functionality)
## Memory requirements
<!--- Compare memory requirements to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

This can be done using https://github.com/pythonprofilers/memory_profiler for CPU memory measurements,
GPU measurements can be done with https://pytorch.org/docs/master/generated/torch.cuda.max_memory_allocated.html.
These tools only profile the memory used by each process, not the entire function.
--->

## Performance
<!--- Compare performance to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

Python has an embedded profiler: https://docs.python.org/3.9/library/profile.html
Again, this will only profile the performance on each process. Printing the results with many processes
may be illegible. It may be easiest to save the output of each to a file.
--->

#### Does this change modify the behaviour of other functions? If so, which?
no

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+    Compute the N-dimensional discrete Fourier Transform.
+
+    This function computes the N-dimensional discrete Fourier Transform over any number of axes in an M-dimensional
+    array by means of the Fast Fourier Transform (FFT). By default, all axes are transformed, with the real transform
```
The comment on real transform along one axis and complex transform on the others does not appear in PyTorch. Is there a difference of the conventions of PyTorch and NumPy?

ClaudiaComito replied:
Interesting, this is GitHub copilot extrapolating out of context. The comment on the real transform belongs in `rfftn`. I've modified the docs.

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+    Notes
+    -----
+    This function requires MPI communication if the input array is transformed along the distribution axis.
+    If the input array is 1-D and distributed, this function copies the entire array on each MPI process! i.e. if the array is very large, you might run out of memory.
```
From my point of view, this is OK. However, I would recommend to open a corresponding issue to work on an alternative specifically for this distributed 1d-case (or does this not make any sense?)

ClaudiaComito replied:
Created issue #1263

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+        also not specified. Repeated indices in `axes` means that the transform over that axis is performed multiple
+        times. (default is None)
+    norm : str, optional
+        Normalization mode: 'forward', 'backward', or 'ortho' (see `numpy.fft` for details). Default is "backward".
```
One could also add a reference to the PyTorch-docs here, because they use the same options for `norm`

ClaudiaComito replied:
I've removed the pointer to [numpy.fft](https://numpy.org/doc/stable/reference/routines.fft.html) for now, however that page does have a lot more infos than the torch.fft page. I think it would be good to mention the numpy page in the docs because it's so exhaustive.

mrfh92 commented on heat/fft/tests/test_fft.py:
```diff
@@ -0,0 +1,235 @@
+        self.assertTrue(y.split == 0)
+        self.assert_array_equal(y, np_y)
+
+        # on GPU, test only on less than 4 processes
```
Is there a reason for not testing on 8 GPUs?

ClaudiaComito replied:
I found what caused the problem on more than 3 GPUs and removed this constraint.

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+    "rfft",
+    "rfft2",
+    "rfftn",
+    # "fftfreq",
```
If there is no special reason for omitting these functions, I would suggest to open a separate issue to add them lateron.

ClaudiaComito replied:
these are implemented now

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+"""Provides a collection of Discrete Fast Fourier Transforms (DFFT) and their inverses."""
+
+import torch
```
I saw the hint "Supports torch.half and torch.chalf on CUDA with GPU Architecture SM53 or greater. However it only supports powers of 2 signal length in every transformed dimensions." in the PyTorch docs on fft/fftn ... does this have any implications for us? Or is this such an exotic case that we dont cover anyway...

ClaudiaComito replied:
Let's keep this in mind when we tackle #94

mrfh92 commented on heat/fft/fft.py:
```diff
@@ -0,0 +1,787 @@
+        output_shape[axis] = nyquist_freq
+    else:
+        output_shape[axis] = n
+    print("DEBUGGING: n, output_shape = ", n, output_shape)
```
here, and at some other places debuggin-info gets printed; we should remove that from the code before merging

ClaudiaComito replied:
You're right, removed

mrfh92 commented on heat/fft/tests/test_fft.py:
```diff
@@ -0,0 +1,235 @@
+        self.assert_array_equal(y, np_y)
+        backwards = ht.fft.irfftn(y, s=x.shape[-2:])
+        self.assertTrue(ht.allclose(backwards, x))
+        # FFT along all axes
```
I have added a comment to that PR in order that we do not forget this

mrfh92 reviewed (commented):
From my point of view this looks fine. Thanks for the work :+1: 
I only have added some minor questions/comment that we can address in the PR talk; after that I guess we can merge. 

In addition, we could already add a benchmark for FFT if this makes sense.

ClaudiaComito commented:
> From my point of view this looks fine. Thanks for the work 👍 I only have added some minor questions/comment that we can address in the PR talk; after that I guess we can merge.
> 
> In addition, we could already add a benchmark for FFT if this makes sense.

@mrfh92 thanks a lot for the review. I think I've addressed all of your comments. I will add the benchmarks next week.

mrfh92 reviewed (approved):
As far as I can see, everything is fine now. 
Thanks for your work! :+1:

mrfh92 commented:
It looks like that we have the same problem with the old PyTorch versions as in #933
